### ISSUE TITLE
should not be able to move the overlay outside the visible screen

### ISSUE THREAD

**Issue by balmas** [alpheios-project/components#472]
I am still able to move the overlay outside the visible screen both in Chrome and Firefox.

_Originally posted by @monzug in https://github.com/alpheios-project/components/issues/342#issuecomment-486676793_

**Comment by balmas** [comment]
this can be tested in 3.0.0.31 and embed-lib 1.0.0.7

**Comment by monzug** [comment]
in this site: http://www.latin.it/autore/agostino/confessiones/!01!liber_i   in chrome
I moved the overlay all the way to the top of the screen. Now the overlay is partially hidden by the screen so I can't move it around. this site has also some css issue (all the text is displayed in the middle not on the left side).
everywhere else the overlay bounced back into the screen

**Comment by kirlat** [comment]
There is a partial fix for some sites ready, but some others require something seemingly much more complex than that. I'm afraid we might need to debug `interact.js` which restricts the toolbar within the specified limits currently. It seems that some smart/not-so-smart layouts may trick it into doing not what it's expected to do.

**Comment by balmas** [comment]
ready for testing in qa-3.0.0.35/1.0.0.13

**Comment by monzug** [comment]
Firefox: latin library, I can move the overlay only once than it freeze. only reloading the page , the overlay can be moved again only once. I confirm that I can still move it outside the screen (on top and bottom) but by reloading the page the overlay is back visible on the screen. also in http://dcc.dickinson.edu/allen-greenough/Chapters/Chapter-100.html. 
Chrome: all good.

**Comment by kirlat** [comment]
I can reproduce issues with both dragging out of the viewport and with not being able to drag once the toolbar is moved on http://dcc.dickinson.edu/allen-greenough/Chapters/Chapter-100.html. I did a lot of tests with different settings and test configurations but, unfortunately, I was not able to find a way to fix it as of today. My conclusion at the moment is that it's something inside `interact.js` that's the reason. Somehow something that's on the page and the way Firefox renders it (because it is fine in Chrome) cause `interact.js` to lose track of a draggable element. `interact.js` loses both the element and the bounds of the page and it all gets messed up.

Once we move mouse over the draggable object, cursor turns into a "move" one. This is done by `interact.js`; we have no code in Alpheios code base that does that. However, after the toolbar is moved on the page listed above in Firefox for the first time, the "move" cursor stops to appear when we move pointer over the draggable element. That's an indication of 'interact.js` stopped working properly.

The same is probably true for viewport bounds. They are calculated and passed to Interact object correctly. Yet it does not work in preventing toolbar from being dragged out of the viewport.

I don't know what fix we can do to make the toolbar being able to be dragged several times without going into long and painful process of debugging an Interact library. We can check manually if element is out of bounds when moved, but it will probably be less effective than what `interact.js` does. And since we need our custom bounds check in Firefox only and on small number of sites, and also because we cannot distinguish the "bad" sites where we need this custom check from the "good" ones, by implementing it we'll be "punishing" majority of our users by providing a less efficient algorithm for them. But still providing our own in-bounds check can be a valid option.

I'm wondering if it's possible to estimate what share of total number of sites might be affected by this issue. Also, it is for Firefox users only and that, given current browser's market share numbers, is approximately one tenth of all the users (maybe more if Firefox is more popular among our user base). But if the number of affected use cases is high enough, we should put an effort into providing our own custom implementation. If the number is low, that we can hope it is a bug in either Firefox or `ineract.js` that will be fixed one day.

What do you think about that?

**Comment by monzug** [comment]
it's firefox only.  Other sites where overlay can be dragged out of the visible area are https://www.tradurreantico.it/latino/grammatica-latina/morfologia/verbo/verbo-eo-latino/  and
https://scaife.perseus.org/reader/urn:cts:latinLit:phi0620.phi001.perseus-lat3:1.2/
I think we should add this in the know issues. by relaoding the page, the overlay comes back except in https://www.tradurreantico.it/latino/grammatica-latina/morfologia/verbo/verbo-eo-latino/

**Comment by balmas** [comment]
> it's firefox only. Other sites where overlay can be dragged out of the visible area are https://www.tradurreantico.it/latino/grammatica-latina/morfologia/verbo/verbo-eo-latino/ and
> https://scaife.perseus.org/reader/urn:cts:latinLit:phi0620.phi001.perseus-lat3:1.2/
> I think we should add this in the know issues. by relaoding the page, the overlay comes back except in https://www.tradurreantico.it/latino/grammatica-latina/morfologia/verbo/verbo-eo-latino/

Ok, added an FAQ label and assigning to myself for adding there.

**Comment by balmas** [comment]
Maybe we lost this fix or something else broke it. I was able to move the overlay outside of the visible area of the screen in the texts-test and texts-beta environment in Chrome (embed-lib 1.0.0.21+)

**Comment by monzug** [comment]
uhm, in Chrome build 3.0.0.39, site https://latin.packhum.org/loc/1017/6/6/1534-1540#6
I was able to move the overlay outside the visible area. on reload, it come back.

**Comment by balmas** [comment]
yeah, something seems to have changed here

**Comment by kirlat** [comment]
I have not had a chance to check it out yet, but the existing solution was not working an all sites. There are sites with such crazy layouts (often with errors in their HTML code) that a built in boundary control of Interact.js (that we're using) failed completely. We did not go with implementing our own functionality for that, and it's not clear if we can do it better than Interact.js. But it was on small number of sites only.

If, however, it stopped working where it was working before, then definitely some of the code is lost, or the layout has changed without corresponding changes to the code.

I will take a look at that after the performance issue (that one is another tricky fellow 🙂).

**Comment by balmas** [comment]
@kirlat just a reminder to take a look at this one too. Thanks!

**Comment by kirlat** [comment]
Will do the the best I can

**Comment by kirlat** [comment]
The fix for this has been merged into master

**Comment by balmas** [comment]
qa-3.0.0.49/components 1.2.12

**Comment by monzug** [comment]
If I move the flex nav outside of the visible screen in all the above sites, it bounces back immediately. in Latin library, it's half way in-half way out, but still usable. the problem is in FF, the overlay (flex nav) then does not move anymore. this has been reported as a separate issue in #712

**Comment by balmas** [comment]
@monzug are the only remaining issues with this the problems on FF reported in #712?

**Comment by monzug** [comment]
yes. I'll close this
